### PR TITLE
Fix emoji bold text

### PR DESCRIPTION
**Pull request by srdjanrist**
### Details
On Google Chrome on Windows render emojis colorless when they have font-weight over 500. 
This update resolves this by wrapping emojis in specific Text components and giving them normal font-weight.

### Fixed Issues

$ [#6772](https://github.com/Expensify/App/issues/6772)

### Tests

1. Open app
2. Check list of reports with unread messages from users that have emojis in their name
4. Verify that their names appear bolded and emojis have color
5. Open specific report and look at messages from the user
6. Verify that users name appear bolded and emojis have color
7. Open users settings 
8. Verify that users name appear bolded and emojis have color

- [ ] Verify that no errors appear in the JS console

### QA Steps

None

- [ ] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web
Mac Chrome
web-mac-chrome
Mac Safari
web-mac-safari
Win Chrome
web-win-chrome

##### Android
android-web-allreports
android-web-messages
android-web-settings

##### iOS
ios-web-allreports
ios-web-messages
ios-web-settings

#### Desktop
desktop

#### iOS
native-ios-allreports
native-ios-messages
native-ios-settings

#### Android
native-android-allreports
native-android-messages
native-android-settings

### CODE REVIEW

**Comment by srdjanrist** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by srdjanrist** [comment]
recheck

**Review by parasharrajat — CHANGES_REQUESTED** [review]
First, @srdjanrist Could you please read the styleGuide and update the PR as much as possible, then I will review it.

File: src/pages/home/report/ReportActionItemFragment.js [hunk 1]
```diff
@@ -93,7 +94,7 @@ class ReportActionItemFragment extends React.PureComponent {
                             numberOfLines={this.props.isSingleLine ? 1 : undefined}
                             style={[styles.chatItemMessageHeaderSender]}
                         >
-                            {Str.htmlDecode(this.props.fragment.text)}
```
[the remark this answers is not in the document]
Comment (srdjanrist) [reply src/pages/home/report/ReportActionItemFragment.js:96]: This was a stupid mistake made by me thinking this decoding wasn't necessary.
Fixed this in the updated code.

[the remark this answers is not in the document]
Comment (srdjanrist) [reply src/pages/home/report/ReportActionItemFragment.js:96]: I have also read the styleGuide and according to them created a utility file. It contains CSS class for fwNormal like Bootstrap does it (https://getbootstrap.com/docs/5.0/utilities/text/#font-weight-and-italics)

**Review by thienlnam — CHANGES_REQUESTED** [review]
It's looking good!
Could you add some unit tests for your new method that involve names with and without emojis please? Something like this - https://github.com/Expensify/App/blob/6435bdf197704c6ae1b36474e867260ddb0dd01d/tests/unit/EmojiRegexTest.js
Just to make sure it doesn't replace unexpected things

File: src/components/EmojiText.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+const propTypes = {
+    /** Text to display */
+    children: PropTypes.string,
+
+    /** Styles to be assigned to Container */
+    style: PropTypes.arrayOf(PropTypes.object),
+};
+
+const defaultProps = {
+    children: null,
```
Comment (parasharrajat) [inline src/components/EmojiText.js:17]:
```suggestion
```

File: src/components/EmojiText.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import Text from './Text';
+import * as EmojiUtils from '../libs/EmojiUtils';
+
```
Comment (parasharrajat) [inline src/components/EmojiText.js:6]: Extra line
```suggestion
```

File: src/components/EmojiText.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    /** Styles to be assigned to Container */
+    style: PropTypes.arrayOf(PropTypes.object),
+};
+
+const defaultProps = {
+    children: null,
+    style: {},
+};
+
+const EmojiText = props => (
```
Comment (parasharrajat) [inline src/components/EmojiText.js:21]: Let's leave a comment on why is this necessary. Why did we have to go through the EmojiText implementation?

File: src/components/EmojiText.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    <Text style={props.style}>{EmojiUtils.escapeEmojiFromText(props.children)}</Text>
```
Comment (parasharrajat) [inline src/components/EmojiText.js:22]: Let's make index.native.js file where `EmojiText` simply extends our `Text` component. And move this logic to index.js and fully extend the `Text` component which means the same prop definition and defaultProps as Text component.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/libs/EmojiUtils.js [hunk 6]
```diff
@@ -155,6 +158,45 @@ function addToFrequentlyUsedEmojis(frequentlyUsedEmojis, newEmoji) {
...
+
+    if (matchedEmojis == null) {
+        return text;
+    }
+
+    const elements = [];
+    let lastIndex = 0;
+    for (let i = 0; i < matchedEmojis.length; i++) {
+        const currentEmoji = matchedEmojis[i];
+        const indexOfEmoji = text.indexOf(currentEmoji, lastIndex);
```
[the remark this answers is not in the document]
Comment (srdjanrist) [reply src/libs/EmojiUtils.js:174]: I have tried to use it, but it didn't yield proper results.

File: src/components/EmojiText.js (hunk 5 above)
Comment (srdjanrist) [reply src/components/EmojiText.js]: Please check latest [commit](https://github.com/Expensify/App/pull/7102/commits/1f21755d3fda60e754313c1bdd1d8e65624c0a58) to see how I have implemented what you have suggested. Hopefully, this is what you intended to do here?

**Comment by srdjanrist** [comment]
> It's looking good! Could you add some unit tests for your new method that involve names with and without emojis please? Something like this - https://github.com/Expensify/App/blob/6435bdf197704c6ae1b36474e867260ddb0dd01d/tests/unit/EmojiRegexTest.js Just to make sure it doesn't replace unexpected things

Thanks for the suggestion! I have updated the pull request with a few tests to check the new method.

File: src/components/EmojiText/index.native.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import Text from '../Text';
+
+// As this workaround is not needed on the native platform, we simply render Text  
+const EmojiText = props => (
+    <Text {...props} />
+);
+
+EmojiText.displayName = 'EmojiText';
+
```
Comment (parasharrajat) [inline src/components/EmojiText/index.native.js:10]: Please add prop definitions that are used for Text component.

File: src/components/EmojiText/index.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    /** Any additional styles to apply */
+    style: stylePropTypes,
+};
+const defaultProps = {
+    color: themeColors.text,
+    fontSize: variables.fontSizeNormal,
+    family: 'GTA',
+    style: {},
+};
```
Comment (parasharrajat) [inline src/components/EmojiText/index.js:35]: import the same from Text component and dry this up.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
a few changes

File: src/components/EmojiText/index.native.js (hunk 7 above)
Comment (srdjanrist) [reply src/components/EmojiText/index.native.js:13]: I have updated the pull request

File: src/components/Text.js [hunk 9]
```diff
@@ -74,3 +74,5 @@ Text.defaultProps = defaultProps;
 Text.displayName = 'Text';
 
 export default Text;
+
+export {propTypes, defaultProps};
```
Comment (parasharrajat) [inline src/components/Text.js:78]: Sorry, but we don't do it this way. 
here is the suggestion
1. Create `baseTextPropTypes.js` file and export {propTypes, defaultProps};. 
2. import as * from baseTextPropTypes on both Text and EmojiText components.
3. Place baseTextPropTypes in Text component folder.

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/Text.js (hunk 9 above)
Comment (srdjanrist) [reply src/components/Text.js]: I have updated the pull request with new commits which takes care of this.
Thank you for your patience.

File: src/components/EmojiText/index.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import Text from '../Text';
+import {propTypes, defaultProps} from '../Text/baseTextPropTypes';
```
Comment (parasharrajat) [inline src/components/EmojiText/index.js:3]:
```suggestion
import * as  baseTextPropTypes from '../Text/baseTextPropTypes';
```

File: src/components/EmojiText/index.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+// This is a workaround on the Chrome browser issue when rendering emojis with font-weight greater than 500
+// More info: https://bugs.chromium.org/p/chromium/issues/detail?id=1266022
+// Text that contains emojis is split so that each emoji is wrapped with a Text component with normal font-weight
+const EmojiText = props => (
+    // eslint-disable-next-line react/jsx-props-no-spreading
+    <Text {...props}>{EmojiUtils.escapeEmojiFromText(props.children)}</Text>
+);
+
+EmojiText.propTypes = propTypes;
+EmojiText.defaultProps = defaultProps;
```
Comment (parasharrajat) [inline src/components/EmojiText/index.js:15]:
```suggestion
EmojiText.propTypes = baseTextPropTypes.propTypes;
EmojiText.defaultProps = baseTextPropTypes.defaultProps;
```

File: src/components/EmojiText/index.native.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import Text from '../Text';
+import {propTypes, defaultProps} from '../Text/baseTextPropTypes';
```
Comment (parasharrajat) [inline src/components/EmojiText/index.native.js:3]:
```suggestion
import * as baseTextPropTypes from '../Text/baseTextPropTypes';
```

File: src/components/EmojiText/index.native.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+// As this workaround is not needed on the native platform, we simply render Text  
+const EmojiText = props => (
+    // eslint-disable-next-line react/jsx-props-no-spreading
+    <Text {...props} />
+);
+
+EmojiText.propTypes = propTypes;
+EmojiText.defaultProps = defaultProps;
```
Comment (parasharrajat) [inline src/components/EmojiText/index.native.js:12]:
```suggestion
EmojiText.propTypes = baseTextPropTypes.propTypes;
EmojiText.defaultProps = baseTextPropTypes.defaultProps;
```

File: src/components/Text/index.js [hunk 14]
```diff
@@ -1,41 +1,11 @@
...
-import PropTypes from 'prop-types';
 import _ from 'underscore';
 // eslint-disable-next-line no-restricted-imports
 import {Text as RNText} from 'react-native';
-import fontFamily from '../styles/fontFamily';
-import themeColors from '../styles/themes/default';
-import variables from '../styles/variables';
+import fontFamily from '../../styles/fontFamily';
+import variables from '../../styles/variables';
+import {propTypes, defaultProps} from './baseTextPropTypes';
```
Comment (parasharrajat) [inline src/components/Text/index.js:7]:
```suggestion
import * as baseTextPropTypes from './baseTextPropTypes';
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]
A couple of style changes. Otherwise, linter will blame

File: src/components/Text/index.js (hunk 14 above)
Comment (srdjanrist) [reply src/components/Text/index.js]: New commit has been pushed with these changes.

**Review by parasharrajat — APPROVED** [review]
LGTM. I didn't test it on Windows. But based on the root cause analysis, I think it will work. 

cc:  @thienlnam 
:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by thienlnam** [comment]
Thanks for the review @parasharrajat, commented in the issue but seems like our QA is not reproducing the issue so if it is already resolved I don't think we need to move forward with this.  We'll still pay both of you for your time but we should probably check that first

**Comment by thienlnam** [comment]
Closing since issue resolved within Chrome
